python/ql/src/semmle/python/objects/ObjectInternal.qll

```diff
@@ -0,0 +1,490 @@
+/**
+ * Internal object API.
+ * For use by points-to and testing only.
+ */
+
+private import semmle.python.objects.TObject
+private import semmle.python.pointsto.PointsTo
+private import semmle.python.pointsto.MRO
+private import semmle.python.pointsto.PointsToContext
+private import semmle.python.types.Builtins
+import semmle.python.objects.Modules
+import semmle.python.objects.Classes
+import semmle.python.objects.Instances
+import semmle.python.objects.Callables
+import semmle.python.objects.Constants
+import semmle.python.objects.Sequences
+import semmle.python.objects.Descriptors
+
+
+class ObjectInternal extends TObject {
+
+    abstract string toString();
+
+    /** The boolean value of this object, this may be both
+     * true and false if the "object" represents a set of possible objects. */
+    abstract boolean booleanValue();
+
+    /** Holds if this object is introduced into the code base at `node` given the `context`
+     * This means that `node`, in `context`, points-to this object, but the object has not flowed
+     * there from anywhere else.
+     * Examples:
+     *  * The object `None` is "introduced" by the keyword "None".
+     *  * A bound method would be "introduced" when relevant attribute on an instance
+     * is accessed. In `x = X(); x.m` `x.m` introduces the bound method.
+     */
+    abstract predicate introduced(ControlFlowNode node, PointsToContext context);
+
+    /** Gets the class declaration for this object, if it is a class with a declaration. */
+    abstract ClassDecl getClassDeclaration();
+
+    /** True if this "object" is a class. That is, its class inherits from `type` */
+    abstract boolean isClass();
+
+    /** Gets the class of this object.  */
+    abstract ObjectInternal getClass();
+
+    /** True if this "object" can be meaningfully analysed to determine its boolean value in comparisons.
+     * For example, `None` or `int` can be, but `int()` or an unknown string cannot.
+     */
+    abstract boolean isComparable();
```

Comment:
"Comparable" seems to be more general than what this actually promises. For instance, I wouldn't call `None` comparable, as e.g. `None < None` doesn't work. Rather, this seems to be about whether the object in question can be meaningfully tested for equality to other objects. `hasEquality`, perhaps?

Reply:
I don't really like `hasEquality` it suggests to me a check to see if the object has a `__eq__` method.
I've used `testableForEquality` instead, and updated the comment.